vendor runc libraryv1.0.0-rc91-48-g67169a9d

tao12345666333 opened the pull request:
<s>full diff https://github.com/opencontainers/runc/compare/v1.0.0-rc10...v1.0.0-rc91</s>

* vendor containerd 779ef60231a555f7eb9ba82b052d59b69ca2ef10 https://github.com/containerd/containerd/compare/c80284d4b5291a351bb471bcdabb5c1d95e7a583...779ef60231a555f7eb9ba82b052d59b69ca2ef10
    - https://github.com/containerd/fifo/compare/ff969a566b00877c63489baf6e8c35d60af6142c...f15a3290365b9d2627d189e619ab4008e0069caf
* vendor runc 67169a9d43456ff0d5ae12b967acb8e366e2f181  https://github.com/opencontainers/runc/compare/v1.0.0-rc10...67169a9d43456ff0d5ae12b967acb8e366e2f181
* vendor bbolt v1.3.5 https://github.com/etcd-io/bbolt/compare/v1.3.3...v1.3.5
    - add go.mod
    - fix panic: runtime error: unsafe pointer conversion (Go 1.14)
* vendor prometheus dependency.
* vendor protobuf v1.3.5 https://github.com/golang/protobuf/compare/v1.3.3...v1.3.5
    - ptypes: regenerate with v3.11.4, update for go1.14 gofmt
    - go.mod: set go version to go1.9

AkihiroSuda reviewed (changes requested):
Please update containerd together

AkihiroSuda commented on vendor.conf:
```diff
@@ -123,7 +123,7 @@ github.com/googleapis/gax-go                        317e0006254c44a0ac427cc52a0e
 
 # containerd
-github.com/containerd/containerd                    c80284d4b5291a351bb471bcdabb5c1d95e7a583 # master / v1.4.0-dev
+github.com/containerd/containerd                    5c695da77b0823df3b4a79100f0fb63f9ce6446b # master / v1.4.0-beta.1-73-g5c695da7
```
the dependencies of containerd should be renvendored as well

tao12345666333 replied:
Thanks. Let me modify it.

tao12345666333 replied:
Addressed.

tao12345666333 commented:
When start the container with `--privileged`, an error will occur, I will check tonight.

thaJeztah commented on daemon/stats/collector_unix.go:
```diff
@@ -8,14 +8,17 @@ import (
+	// on Linux it's a constant which is safe to be hard coded,
+	// so we can avoid using cgo here. For details, see:
+	// https://github.com/containerd/cgroups/pull/12
+	s.clockTicksPerSecond = uint64(100)
```
Perhaps this should now just be a `const clockTicksPerSecond` (grouped with the `nanoSecondsPerSecond` below), and then we can remove the `clockTicksPerSecond` property, and `platformNewStatsCollector()`.

I think this change would make sense as a separate commit (or even a separate PR)

cpuguy83 reviewed (dismissed):
LGTM

cpuguy83 commented:
Lots of failures, though. :(

tao12345666333 commented:
The error is now at https://github.com/tao12345666333/moby/blob/update-runc-to-v1.0.0-rc91/libcontainerd/remote/client.go#L247

thaJeztah commented:
Opened https://github.com/moby/moby/pull/41186 to address https://github.com/moby/moby/pull/41178/files#r450124125

tao12345666333 commented:
I found that the final error message comes from
https://github.com/opencontainers/runc/blob/v1.0.0-rc91/libcontainer/rootfs_linux.go#L654-L656


The `unix.Mknod` method returned an  `invalid argument` error.

I checked the value it got, e.g. 

dest:`/var/lib/docker/overlay2/f91757214cfde0ac98dbb16bc3d24e358fbea3f7cf47914ebc1f90d539687be6/merged/dev/core` 
fileMode: `-rwxrwxrwx`
dev: `0`

node: `{"type":99,"major":0,
"minor":0,"permissions":"","allow":false,"path":"/dev/core","file_mode":41471,"uid":0,"gid":0}`

Checking.

thaJeztah commented:
I think we need https://github.com/moby/moby/pull/41016 as well

tao12345666333 commented:
> I think we need #41016 as well

Thanks. Let me take a look.

tao12345666333 commented:
> > I think we need #41016 as well
> 
> Thanks. Let me take a look.

I tried to merge it in and test it locally. But the error encountered here still exists.  
I need to keep checking this issue.

thaJeztah commented:
https://github.com/moby/moby/pull/41016 was merged; could you rebase the PR (to check if it helps with the failures)

tao12345666333 commented:
sure!

tao12345666333 commented:
The problem still exists, when the `--privileged` is passed, the container cannot be started normally.

Even if I upgrade containerd to v1.4.0 (#40982 ), the problem is the same.
I will spend time next week to continue this/

AkihiroSuda commented:
needs rebase

tao12345666333 commented:
When I used `strace` to debug it, I found an unexpected behavior:

```
[pid 21759] 04:15:13 mknodat(AT_FDCWD, "/var/lib/docker/overlay2/ec5e3eeca5c20f09d0213a51d69c34b1ce4ea4fa3421c08d2ef86c7cba353611/merged/dev/core", S_IFLNK|0777) = -1 EINVAL (Invalid argument)           
```

**The argument is `S_IFLNK`. **   (Not finished editing, add later).


Full context here. 
```
[pid 21760] 04:15:13 nanosleep({tv_sec=0, tv_nsec=20000},  <unfinished ...>                                                            
[pid 21759] 04:15:13 <... mkdirat resumed> ) = 0                                                                                       
[pid 21759] 04:15:13 mknodat(AT_FDCWD, "/var/lib/docker/overlay2/ec5e3eeca5c20f09d0213a51d69c34b1ce4ea4fa3421c08d2ef86c7cba353611/merged/dev/bus/usb/002/001", S_IFCHR|0664, makedev(0xbd, 0x80) <unfinished ...>                                                             
[pid 21760] 04:15:13 <... nanosleep resumed> NULL) = 0                                                                                 
[pid 21759] 04:15:13 <... mknodat resumed> ) = 0                                                                                       
[pid 21760] 04:15:13 nanosleep({tv_sec=0, tv_nsec=20000},  <unfinished ...>                                                            
[pid 21759] 04:15:13 fchownat(AT_FDCWD, "/var/lib/docker/overlay2/ec5e3eeca5c20f09d0213a51d69c34b1ce4ea4fa3421c08d2ef86c7cba353611/merged/dev/bus/usb/002/001", 0, 0, 0) = 0                                                                                                  
[pid 21759] 04:15:13 newfstatat(AT_FDCWD, "/var/lib/docker/overlay2/ec5e3eeca5c20f09d0213a51d69c34b1ce4ea4fa3421c08d2ef86c7cba353611/merged/dev",  <unfinished ...>                                                                                                           
[pid 21760] 04:15:13 <... nanosleep resumed> NULL) = 0                                                                                 
[pid 21759] 04:15:13 <... newfstatat resumed> {st_mode=S_IFDIR|0755, st_size=300, ...}, 0) = 0                                         
[pid 21760] 04:15:13 nanosleep({tv_sec=0, tv_nsec=20000},  <unfinished ...>                                                            
[pid 21759] 04:15:13 mknodat(AT_FDCWD, "/var/lib/docker/overlay2/ec5e3eeca5c20f09d0213a51d69c34b1ce4ea4fa3421c08d2ef86c7cba353611/merged/dev/core", S_IFLNK|0777) = -1 EINVAL (Invalid argument)                                                                              
[pid 21759] 04:15:13 umask(022)         = 000                                                                                          
[pid 21760] 04:15:13 <... nanosleep resumed> NULL) = 0                                                                                 
[pid 21759] 04:15:13 write(4, "{\"type\":\"procError\"}", 20 <unfinished ...>                                                          
[pid 21760] 04:15:13 nanosleep({tv_sec=0, tv_nsec=20000},  <unfinished ...>                                                            
[pid 21759] 04:15:13 <... write resumed> ) = 20                                                                                        
[pid 21760] 04:15:13 <... nanosleep resumed> NULL) = 0                                                                                 
[pid 21756] 04:15:13 <... read resumed> "{\"type\":\"procError\"}", 512) = 20                                                          
[pid 21760] 04:15:13 nanosleep({tv_sec=0, tv_nsec=20000},  <unfinished ...>                                                            
[pid 21756] 04:15:13 futex(0xdfb110, FUTEX_WAKE_PRIVATE, 1) = 1                                                                        
[pid 21751] 04:15:13 <... futex resumed> ) = 0                                                                                         
[pid 21759] 04:15:13 openat(AT_FDCWD, "/etc//localtime", O_RDONLY <unfinished ...>
[pid 21751] 04:15:13 nanosleep({tv_sec=0, tv_nsec=20000},  <unfinished ...>
```

tao12345666333 commented:
rebased.

thaJeztah commented:
Seeing these failures;

```
=== RUN   TestDockerSuite/TestBuildRUNoneJSON
    --- FAIL: TestDockerSuite/TestBuildRUNoneJSON (0.18s)
        docker_cli_build_test.go:3957: assertion failed:
            Command:  /usr/local/cli/docker build -t testbuildrunonejson -
            ExitCode: 1
            Error:    exit status 1
            Stdout:   Sending build context to Docker daemon  2.048kB

            Step 1/2 : FROM hello-world:frozen
             ---> bf756fb1ae65
            Step 2/2 : RUN [ "/hello" ]
             ---> Running in e63c05e10db8

            Stderr:   unable to find user 0: invalid argument


            Failures:
            ExitCode was 1 expected 0
            Expected stdout to contain "Hello from Docker"
            Expected no error
```

Looks like they are introduced in https://github.com/opencontainers/runc/commit/0af5cd204103a23b047bd4b8e4159f8ef5a1ef3c (https://github.com/opencontainers/runc/pull/2275)

I tried to apply only those changes (using moby code from master); and with this patch applied to the vendor directory

<details>

```patch
diff --git a/vendor/github.com/opencontainers/runc/libcontainer/user/user.go b/vendor/github.com/opencontainers/runc/libcontainer/user/user.go
index 7b912bbf8b..de30982ba2 100644
--- a/vendor/github.com/opencontainers/runc/libcontainer/user/user.go
+++ b/vendor/github.com/opencontainers/runc/libcontainer/user/user.go
@@ -162,10 +162,6 @@ func ParsePasswdFilter(r io.Reader, filter func(User) bool) ([]User, error) {
 	)
 
 	for s.Scan() {
-		if err := s.Err(); err != nil {
-			return nil, err
-		}
-
 		line := strings.TrimSpace(s.Text())
 		if line == "" {
 			continue
@@ -183,6 +179,9 @@ func ParsePasswdFilter(r io.Reader, filter func(User) bool) ([]User, error) {
 			out = append(out, p)
 		}
 	}
+	if err := s.Err(); err != nil {
+		return nil, err
+	}
 
 	return out, nil
 }
@@ -221,10 +220,6 @@ func ParseGroupFilter(r io.Reader, filter func(Group) bool) ([]Group, error) {
 	)
 
 	for s.Scan() {
-		if err := s.Err(); err != nil {
-			return nil, err
-		}
-
 		text := s.Text()
 		if text == "" {
 			continue
@@ -242,6 +237,9 @@ func ParseGroupFilter(r io.Reader, filter func(Group) bool) ([]Group, error) {
 			out = append(out, p)
 		}
 	}
+	if err := s.Err(); err != nil {
+		return nil, err
+	}
 
 	return out, nil
 }
@@ -532,10 +530,6 @@ func ParseSubIDFilter(r io.Reader, filter func(SubID) bool) ([]SubID, error) {
 	)
 
 	for s.Scan() {
-		if err := s.Err(); err != nil {
-			return nil, err
-		}
-
 		line := strings.TrimSpace(s.Text())
 		if line == "" {
 			continue
@@ -549,6 +543,9 @@ func ParseSubIDFilter(r io.Reader, filter func(SubID) bool) ([]SubID, error) {
 			out = append(out, p)
 		}
 	}
+	if err := s.Err(); err != nil {
+		return nil, err
+	}
 
 	return out, nil
 }
@@ -586,10 +583,6 @@ func ParseIDMapFilter(r io.Reader, filter func(IDMap) bool) ([]IDMap, error) {
 	)
 
 	for s.Scan() {
-		if err := s.Err(); err != nil {
-			return nil, err
-		}
-
 		line := strings.TrimSpace(s.Text())
 		if line == "" {
 			continue
@@ -603,6 +596,9 @@ func ParseIDMapFilter(r io.Reader, filter func(IDMap) bool) ([]IDMap, error) {
 			out = append(out, p)
 		}
 	}
+	if err := s.Err(); err != nil {
+		return nil, err
+	}
 
 	return out, nil
 }
```

</details>


And with that patch applied, I see the same error;

```
DOCKER_BUILDKIT=0 docker build -<<EOF
> FROM hello-world
> RUN [ "/hello" ]
> EOF
Sending build context to Docker daemon  2.048kB
Step 1/2 : FROM hello-world
latest: Pulling from library/hello-world
0e03bdcc26d7: Pull complete
Digest: sha256:49a1c8800c94df04e9658809b006fd8a686cab8028d33cfba2cc049724254202
Status: Downloaded newer image for hello-world:latest
 ---> bf756fb1ae65
Step 2/2 : RUN [ "/hello" ]
 ---> Running in c88e29df3526
unable to find user 0: invalid argument
```

@kolyshkin PTAL

tao12345666333 commented:
Continue to debug and find that the problem (Invalid argument error when start privileged container) is introduced in [opencontainers/runc#2391](https://github.com/opencontainers/runc/pull/2391)

This is related to the change in the device part, but I did not locate which commit it came from.  @cyphar  PTAL. Thanks!

cyphar commented:
I think I found the issue, `DeviceFromPath` (which is used by Docker via `HostDevices`) isn't correctly handling `S_IFMT` checks (and the PR you've linked just uncovered this bug). Basically the issue boils down to the `S_IFIFO` and other file-type flags not being bitflags -- you need to use `&S_IFMT == S_IFWHATEVER` in order to correctly check the type of the file. As a result, a symlink ends up being counted as a character device (`S_IFLNK & S_IFCHR != 0`).

opencontainers/runc#2529 should fix the issue.

tao12345666333 commented:
Thanks for your quick response @cyphar 
I will test it ASAP.

tao12345666333 commented:
When I apply the patch (https://github.com/opencontainers/runc/pull/2529) locally, privileged container can start. :+1:

tao12345666333 commented:
The test case `TestBuildRUNoneJSON` passed.

```
[2020-07-29T04:32:56.248Z]     --- PASS: TestDockerSuite/TestBuildRUNErrMsg (0.51s)
[2020-07-29T04:32:56.248Z]     --- PASS: TestDockerSuite/TestBuildRUNoneJSON (0.57s)
[2020-07-29T04:32:56.248Z]     --- PASS: TestDockerSuite/TestBuildRelativeCopy (5.26s)
```

Thanks @cyphar  @thaJeztah

thaJeztah commented:
Let me take that last commit and open it as a separate PR, as that looks like an existing bug

thaJeztah commented:
opened https://github.com/moby/moby/pull/41288

tao12345666333 commented:
👌 Thanks

thaJeztah commented on vendor.conf:
```diff
@@ -83,8 +83,9 @@ google.golang.org/grpc                              f495f5b15ae7ccda3b38c53a1bfc
-github.com/opencontainers/runtime-spec              c4ee7d12c742ffe806cd9350b6af3b4b19faed6f # v1.0.2
+github.com/opencontainers/runc                      24a3cf88a7ae5f4995f6750654c0e2ca61ef4bb2 # v1.0.0-rc91
+github.com/opencontainers/runtime-spec              237cc4f519e2e8f9b235bacccfa8ef5a84df2875 # v1.0.3-0.20200520003142-237cc4f519e2
+
```
Small nit; on your next rebase/push can you remove this blank line?

tao12345666333 replied:
sure!

tao12345666333 committed:
vendor containerd 779ef60231a555f7eb9ba82b052d59b69ca2ef10

v1.4.0-beta.1-150-g779ef602

Signed-off-by: Jintao Zhang <zhangjintao9020@gmail.com>

tao12345666333 committed:
vendor runc 67169a9d43456ff0d5ae12b967acb8e366e2f181

v1.0.0-rc91-48-g67169a9d

Signed-off-by: Jintao Zhang <zhangjintao9020@gmail.com>

tao12345666333 commented:
CI passed :+1:

thaJeztah reviewed (commented):
left some comments on other dependencies; I can try having a look at opening separate PR's for those (in case we want to cherry-pick those to the 19.03 release branch)

AkihiroSuda commented on vendor/github.com/containerd/cgroups/utils.go:
```diff
@@ -366,5 +382,18 @@ func cleanPath(path string) string {
+	return path
+}
+
+func retryingWriteFile(path string, data []byte, mode os.FileMode) error {
+	// Retry writes on EINTR; see:
+	//    https://github.com/golang/go/issues/38033
```
let's just merge https://github.com/moby/moby/pull/41253 and we will have the latest containerd/cgroups

thaJeztah replied:
Ah, yes, forgot that one was updating it; merged it

tao12345666333 commented:
thanks！ let me update it

tao12345666333 committed:
vendor bbolt v1.3.5

Signed-off-by: Jintao Zhang <zhangjintao9020@gmail.com>

tao12345666333 commented on vendor/github.com/containerd/containerd/vendor.conf:
```diff
@@ -1,102 +1,102 @@
-go.etcd.io/bbolt                                    a0458a2b35708eef59eb5f620ceb3cd1c01a824d # v1.3.3
-go.opencensus.io                                    9c377598961b706d1542bd2d84d538b5094d596e # v0.22.0
+github.com/urfave/cli                               v1.22.1 # NOTE: urfave/cli must be <= v1.22.1 due to a regression: https://github.com/urfave/cli/issues/1092
+go.etcd.io/bbolt                                    v1.3.5
```
OK, I create a separate commit. (Will push soon)

tao12345666333 commented on vendor/github.com/containerd/containerd/vendor.conf:
```diff
@@ -1,102 +1,102 @@
-go.opencensus.io                                    9c377598961b706d1542bd2d84d538b5094d596e # v0.22.0
+github.com/urfave/cli                               v1.22.1 # NOTE: urfave/cli must be <= v1.22.1 due to a regression: https://github.com/urfave/cli/issues/1092
+go.etcd.io/bbolt                                    v1.3.5
+go.opencensus.io                                    v0.22.0
```
So shall I skip it for now?

tao12345666333 committed:
vendor prometheus dependency.

github.com/prometheus/client_golang to v1.6.0
github.com/prometheus/client_model to v0.2.0
github.com/prometheus/common to v0.9.1
github.com/prometheus/procfs to v0.0.11

Signed-off-by: Jintao Zhang <zhangjintao9020@gmail.com>

tao12345666333 committed:
vendor protobuf v1.3.5

Signed-off-by: Jintao Zhang <zhangjintao9020@gmail.com>

tao12345666333 commented on vendor/github.com/containerd/containerd/vendor.conf:
```diff
@@ -1,102 +1,102 @@
+github.com/godbus/dbus/v5                           v5.0.3
+github.com/gogo/googleapis                          v1.3.2
+github.com/gogo/protobuf                            v1.3.1
+github.com/golang/protobuf                          v1.3.5
```
yes. I created a separate commit for update it to v1.3.5.

tao12345666333 commented:
@thaJeztah  PTAL. Thanks

thaJeztah commented on vendor/github.com/containerd/containerd/cio/io_unix.go:
```diff
@@ -132,7 +132,7 @@ func openFifos(ctx context.Context, fifos *FIFOSet) (pipes, error) {
 		}()
 	}
-	if fifos.Stderr != "" {
+	if !fifos.Terminal && fifos.Stderr != "" {
```
I see this fixes a potential panic; https://github.com/containerd/containerd/pull/3611, but I think that was related to a change that's only in 1.3.x, so I guess no need to get that into the 19.03 branch

tao12345666333 replied:
:+1:

tao12345666333 replied:
:+1:

thaJeztah reviewed (approved):
LGTM, thanks!

thaJeztah commented:
Added "cherry-pick" label to consider backporting some of the dependency bumps related to Go 1.14 fixes
